feat(lsp): initial support for dynamic capabilities

folke opened the pull request:
Added initial support for [dynamic registration of capabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#client_registerCapability)

- `client.dynamic_capabilities` is an object that tracks client register/unregister
- `client.supports_method` will additionally check if a dynamic capability supports the method, taking document filters into account. But **only** if the client enabled `dynamicRegistration` for the capability
- updated the default client capabilities to include dynamicRegistration for:
  - `formatting`
  - `rangeFormatting`
  - `hover`
  - `codeAction`
  - `hover`
  - `rename`

## Notes

- `client.server_capabilities` only contains the capabilities set during server initialization.
- Client code that wants to fully support dynamic capabilities should additionally check `client.dynamic_capabities:get(method, {bufnr=...})`
- Maybe I should add a `client.get_server_capability(method, {bufnr=...})` that combines both like with `client.supports_method`?
- I didnt enable `dynamicRegistration` by default for `completion` since that ***possibly*** requires extra processing of `registrationOptions` (see the point above)

## Examples

This opens up a lot of new functionality for existing lsp servers. Some examples:

- formatting now works with rome lsp, since they only advertise formatting with dynamic registration
- eslintls also has a formatter, which does the same as EslintFixAll, so that special handling in lspconfig would no longer be needed
- some lsps advertise additional code actions

Would be great to get some feedback if this is something we would like to have in core.

mfussenegger commented on runtime/lua/vim/lsp/dynamic.lua:
```diff
@@ -0,0 +1,115 @@
+end
+
+--- @param method string
+--- @param opts? {bufnr?: number}
+function M:get(method, opts)
```
Can you add `@return` too?

mfussenegger commented on runtime/lua/vim/lsp.lua:
```diff
@@ -6,6 +6,7 @@ local protocol = require('vim.lsp.protocol')
 local util = require('vim.lsp.util')
 local sync = require('vim.lsp.sync')
 local semantic_tokens = require('vim.lsp.semantic_tokens')
+local dynamic = require('vim.lsp.dynamic')
```
I assume we want people to go through client.dynamic_capabilities, and not access vim.lsp.dynamic directly, so could you rename `dynamic.lua` to `_dynamic` and move the require call to `dynamic_capabilities = require("vim.lsp._dynamic").new()` to make it lazy?

mfussenegger reviewed (commented):
> if this is something we would like to have in core.

Yes we do.

> Maybe I should add a client.get_server_capability(method, {bufnr=...}) that combines both like with client.supports_method?

I'd stick to `client.supports_method` and not expose additional convenience functions that don't enable new use cases to keep the API surface smaller (and hence less complex)

Left a couple of comments, overall the approach looks good to me.

folke committed:
feat(lsp): added dynamic registration of lsp capabilities

folke commented on runtime/lua/vim/lsp/dynamic.lua:
```diff
@@ -0,0 +1,115 @@
+    if not self.capabilities[method] then
+      self.capabilities[method] = {}
+    end
+    table.insert(self.capabilities[method], reg)
```
Not sure about this one. Sounds a bit too magical to me for this use-case :)
We also do checks on the table if a capability was registered and I assume `vim.defaulttable` would then make an entry in that table?

folke commented:
What should we do with `completion`? We can enable dynamic registration for this as well, but some lsp servers I tested with provide additional options.

The current deafult omnifunc implementation also doesn't take dynamic registration of completion into account.

Just leave it as is for now? Can be changed in another PR if it would be useful.

mfussenegger commented:
> Just leave it as is for now? Can be changed in another PR if it would be useful.

Yes let's leave it out for now

folke commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1299,7 +1300,12 @@ function lsp.start_client(config)
       trace = valid_traces[config.trace] or 'off',
+      ---@param method string
     }
+    client.supports_dynamic_registration = function(method)
```
Added this function so we can check if the client supports dynamic registration for a certain method.

I've also updated the register/unregister handlers to only log when an unsupported registration happens. The unsupported methods are also added to the log

folke replied:
Since the initial client capabilities added when starting the client are not exposed anywhere, it's currently not possible to enable dynamicRegistration after the facts.

The registrations will be tracked, but `client.supports_method` will only check those that were enabled during start.

We could expose `client.client_capabilities`, but I would leave it as is for now

mfussenegger replied:
I'd prefer inlining this into `supports_method` instead of exposing an additional method. Users could access `client.config.capabilities` if they need the information.

folke replied:
done, added the function on `_dynamic` instead. Indeed not need to expose on client.

folke commented:
Also worth mentioning that some LSP servers like eslint, where people typically have `EslintFixALL` in a `FileWritePost`, will now also do the same thing as part of format.

Or people that are using rome lsp, together with the null-ls rome formatter (which is no longer needed with this PR)

Not a big deal, and easy to fix in their configs, but something to be aware of.

mfussenegger reviewed (commented):
Could you also add some tests?

folke commented:
@mfussenegger I was just doing that actually. Totally forgot we have capabilities exposed on client.config

folke commented:
@mfussenegger added some tests

folke commented:
Are there any other changes needed before merging?

mfussenegger reviewed (commented):
A side effect of this change is that some servers will no longer report the formatting capabilities in `server_capabilities`, which breaks setting the formatexpr as default in:

https://github.com/folke/neovim/blob/23ce849969c88c9ddde5ec01ff2f30a3593e769e/runtime/lua/vim/lsp.lua#L1125-L1131

I'm not sure how to best handle that. I'm not sure if updating the check to `supports_method` is enough, as the dynamic capabilities may not be there yet when we try to set the defaults?

But we should at least be more verbose about the implications in the news entry.

mfussenegger commented:
On a quick test with eclipse.jdt.ls the codeActionProvider also disappears which means https://github.com/folke/neovim/blob/23ce849969c88c9ddde5ec01ff2f30a3593e769e/runtime/lua/vim/lsp/buf.lua#L686-L689 needs an update as well.

I think the other places where we use the server_capabilities directly should be okay.

folke commented:
That's indeed the same as what I mentioned with completion. I can make some changes so that formatting also supports dynamic registration. By using supports_method and by adding a dynamic formatting handler possibly. Need to double check the code to see that we'd only be overriding defaults and not custom formatter that may have been set by the user in the meantime.

folke committed:
feat(lsp): enabled dynamic registration for textDocument/definition

folke committed:
fix(lsp): make codaAction/resolve work with dynamic registration

folke committed:
fix(lsp): make sure client.supports_method is available during lsp init

folke committed:
fix(lsp): support dynamic registration for lsp defaults (formatpgr etc)

folke commented:
* Enabled dynamic registration for `textDocument/definition`
* moved `set_defaults` and `is_empty_or_default` out of `client.start`
* move defnition of `client.supports_method` out of after initialize
* expose `vim.lsp._set_defaults`
* when an lsp does a dynamic registration, we set defaults again for attached buffers (will only change anything if it's still the default)
* Fixed dynamic registration for `textDocument/resolve` (code actions)

Note: with the lsp's I tested, most dynamic registration happens during lsp inititalize, so that's why I had to move out some methods and define them earlier in the startup process.

folke committed:
fix(lsp): added codeAction/resolve to request_name_to_capability map

folke commented:
I hadn't seen that the tests were failing. I forgot to add an entry for code action resolve to that mapping table.

All good now!

mfussenegger commented on runtime/doc/news.txt:
```diff
@@ -36,6 +36,8 @@ ADDED FEATURES                                                     *news-added*
 
 The following new APIs or features were added.
 
+• Dynamic registration of LSP capabilities
+
```
```suggestion
• Dynamic registration of LSP capabilities. An implication of this change is that checking a client's `server_capabilities` is no longer a sufficient indicator to see if a server supports a feature. Instead use `client.supports_method(<method>)`. It considers both the dynamic capabilities and static `server_capabilities`.
```

folke committed:
docs: update runtime/doc/news.txt

Co-authored-by: Mathias Fußenegger <mfussenegger@users.noreply.github.com>

folke commented:
Updated the docs. Do you want me to put everything in one commit? Or will you just squash/merge it?

mfussenegger commented:
> Do you want me to put everything in one commit? Or will you just squash/merge it?

I'll just squash unless you want me to keep the commits?

dsully commented:
While I love the dynamic registration support, using `client.supports_method("text")` no longer allows for completion of the valid fields, and instead introduces arbitrary text, for which typos/invalid LSP functionality is not easily caught.

folke commented:
@dsully we can fix that in another PR. There's currently a lot of places in the Neovim code where code isn't typed correctly. Shouldn't be a blocker for this.

folke commented:
> > Do you want me to put everything in one commit? Or will you just squash/merge it?
> 
> I'll just squash unless you want me to keep the commits?

nope, squashing is fine. Thanks!

dsully commented:
Thank you!

mfussenegger commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -683,11 +683,7 @@ local function on_code_action_results(results, ctx, options)
-      and client
-      and vim.tbl_get(client.server_capabilities, 'codeActionProvider', 'resolveProvider')
-    then
+    if not action.edit and client and client.supports_method('codeAction/resolve') then
```
Turns out this change doesn't work.
`client.supports_method('codeAction/resolve')` returns `false` if it's supported via dynamic capabilities because there won't be an entry for `codeAction/resolve`, but instead it looks like this:

```
        ["textDocument/codeAction"] = { {
            id = "3b2131ee-bf17-4a3d-a528-ad698f162956",
            method = "textDocument/codeAction",
            registerOptions = {
              codeActionKinds = { "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor
.rewrite", "source", "source.organizeImports" },
              resolveProvider = true
            }
          } },
```

Using code actions which require resolving the command is currently broken on master because of this.

folke replied:
See https://github.com/neovim/neovim/pull/23826

fsouza commented on runtime/doc/news.txt:
```diff
@@ -36,6 +36,7 @@ ADDED FEATURES                                                     *news-added*
 
 The following new APIs or features were added.
 
+• Dynamic registration of LSP capabilities. An implication of this change is that checking a client's `server_capabilities` is no longer a sufficient indicator to see if a server supports a feature. Instead use `client.supports_method(<method>)`. It considers both the dynamic capabilities and static `server_capabilities`.
```
Thank you very much for working on this, it's really exciting! It would be good to have a good alternative to dynamically react to server capabilities though.

For example, in order to conditionally add a mapping for `textDocument/hover`, I used to check server_capabilities in the `LspAttach` callback, and now I have to override the handler for `client/registerCapability`, and we actually need to do both to account for servers that don't support dynamic registration. Basically, this will lead to some duplication between LspAttach and the handler for `client/registerCapability`, similar to what you had to do for `lsp._set_defaults`.

Personally, I think the duplication isn't too bad (can just create a shared function and call from both places), what's annoying is having to override the handler. It works, but I wonder if you have any plans/ideas on how to improve it?

Thank you again!
